torch/csrc/utils/tensor_new.cpp

```diff
@@ -1163,8 +1163,11 @@ Tensor asarray(
     // Make tensor from sequence, inferring its type, and then convert
     // it to the desired type.
     tensor = internal_new_from_data(
-        TensorOptions(), dtype_unwrapped, device, obj, false, false, true);
-    tensor = tensor.to(dtype_unwrapped);
+        TensorOptions(), dtype_unwrapped, device, obj,
+        /* copy_variables = */ false, /* copy_numpy = */ false, /* type_inference = */ true);
+    if (dtype.has_value()) {
+      tensor = tensor.to(dtype_unwrapped);
+    }
```

Comment:
Could this be done by passing hte `dtype` within `TensorOptions()` to avoid the extra copy?

Reply:
I don't think so. `internal_new_from_data` does not use its dtype (maybe it could swallow the dtype parameter in another PR, though). That said, we can make that extra copy go away by `type_inference = !dvalue.has_value()`.

Reply:
I was talking about the extra copy when calling `to`, but I guess you have a point as well with that.

So, doing `type_inference = !dvalue.has_value()` you would not need the `if`  and the `to`, right?
That way we would go from (potentially) 2 copies to just one, which is neat.

Reply:
Yes, exactly!